### PR TITLE
Introduce setting limits for resources (CPU and RAM)

### PR DESCRIPTION
**Pull request by aprattes**
*) Introducing MaxRAM and MaxCPU to limit the resources a framework can
   consume.
*) Fix wrong request might be aborted when buildable item has left the
   queue

Issue: INFRABRBA-3897

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/mesos/monitoring/MesosTaskFailureMonitor.java [hunk 1]
```diff
@@ -89,23 +88,30 @@ private void removeExistingNode(Mesos.JenkinsSlave failedSlave, PrintStream logg
...
-  public boolean addFailedSlave(Mesos.JenkinsSlave slave) {
-    return failedSlaves.add(slave);
+  public void addFailedSlave(Mesos.JenkinsSlave slave, Mesos.SlaveResult.FAILED_CAUSE cause) {
+    failedSlaves.put(slave, cause);
   }
 
-  public Set<Mesos.JenkinsSlave> getFailedSlaves() {
-    return Collections.unmodifiableSet(failedSlaves);
+  public Map<Mesos.JenkinsSlave, Mesos.SlaveResult.FAILED_CAUSE> getFailedSlaves() {
+    return failedSlaves;
```
Comment (ederst) [inline src/main/java/org/jenkinsci/plugins/mesos/monitoring/MesosTaskFailureMonitor.java:115]: Why not `Collections.unmodifiableMap(failedSlaves)`? The map should not be modifiable outside of this class (danger of unnoticed changes, etc.).

File: src/main/java/org/jenkinsci/plugins/mesos/monitoring/MesosTaskFailureMonitor.java [hunk 2]
```diff
@@ -89,23 +88,30 @@ private void removeExistingNode(Mesos.JenkinsSlave failedSlave, PrintStream logg
...
-  public HttpResponse doDismiss() {
+  public void fixTasks() {
+    start(false);
+  }
+
+  public void ignoreTasks() {
     getLogFile().delete();
-    failedSlaves = Collections.synchronizedSet(new LinkedHashSet<Mesos.JenkinsSlave>());
-    return HttpResponses.forwardToPreviousPage();
+    failedSlaves = new HashMap<Mesos.JenkinsSlave, Mesos.SlaveResult.FAILED_CAUSE>();
```
Comment (ederst) [inline src/main/java/org/jenkinsci/plugins/mesos/monitoring/MesosTaskFailureMonitor.java:107]: `HashMap` is not `synchronized` to my knowledge, and the methods changing the contents of the map are neither (`this.failedSlaves.remove()` above, and `failedSlaves.put()` further down), so to my understanding it is possible that an update is missed (even though the variable is volatile)?

File: src/main/java/org/jenkinsci/plugins/mesos/monitoring/MesosTaskFailureMonitor.java [hunk 3]
```diff
@@ -42,13 +40,14 @@ public boolean isActivated() {
   @Override
   public void fix(TaskListener taskListener) throws Exception {
     PrintStream logger = taskListener.getLogger();
-    Set<Mesos.JenkinsSlave> failedSlaves = getFailedSlaves();
-    for (Mesos.JenkinsSlave failedSlave : failedSlaves) {
+
+    Map<Mesos.JenkinsSlave, Mesos.SlaveResult.FAILED_CAUSE> map = new HashMap<Mesos.JenkinsSlave, Mesos.SlaveResult.FAILED_CAUSE>(failedSlaves);
+
+    for (Mesos.JenkinsSlave failedSlave : map.keySet()) {
+      this.failedSlaves.remove(failedSlave);
```
Comment (ederst) [inline src/main/java/org/jenkinsci/plugins/mesos/monitoring/MesosTaskFailureMonitor.java:47]: Possible `ConcurrentModificationException`? --> nope, missed the duplication of the map

Is it necessary to use duplicated map or use delete [via iterator](http://stackoverflow.com/a/1884916/7744760) or [fancy Java 8 code](http://stackoverflow.com/a/22679166/7744760)?

Or, due to the fact that every processed failed slave gets deleted anyways, why is the failedSlaves map not duplicated and the "old" one freshly initialized anyways? (probably with optional .addAll in a finally block)

File: src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/config.jelly [hunk 4]
```diff
@@ -19,6 +19,14 @@
         <f:textbox field="role" default="*" clazz="required"/>
     </f:entry>
 
+    <f:entry title="${%Maximal Cpus allocation}" field="maxCpus">
```
Comment (ederst) [inline src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/config.jelly:22]: Change to "Maximum CPU allocation"

File: src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/config.jelly [hunk 5]
```diff
@@ -19,6 +19,14 @@
...
+        <f:textbox field="maxCpus" default="-1" clazz="required"/>
+    </f:entry>
+
+    <f:entry title="${%Maximal Memory allocation}" field="maxMem">
```
Comment (ederst) [inline src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/config.jelly:26]: Change to "Maximum memory allocation"

File: src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/help-maxMem.html [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<div>
+    The maximum number of memory(RAM) in MB this framework can use.
```
Comment (ederst) [inline src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/help-maxMem.html:2]: Add space "... memory (RAM) in MB..."

**Review by rekie — APPROVED** [review]
thx

**Review by ederst — COMMENTED** [review]
test

**Pull request merged**